File: go/leetcode_easy/max_sub_array/max_sub_array.go
package max_sub_array

func MaxSubArray(nums []int) int {
	bestValue := nums[0]
	previousBest := nums[0]

	for i := 1; i < len(nums); i++ {
		currentValue := nums[i]

		if currentValue < currentValue+previousBest {
			previousBest = currentValue + previousBest
		} else {
			previousBest = currentValue
		}

		if previousBest > bestValue {
			bestValue = previousBest
		}
	}

	return bestValue
}

File: py/leetcode_easy/max_sub_array.py
def maxSubArray(nums):
    bestValue = nums[0]
    previousBest = nums[0]

    i = 1
    while i < len(nums):
        currentValue = nums[i]

        if currentValue < currentValue + previousBest:
            previousBest = currentValue + previousBest
        else:
            previousBest = currentValue

        if previousBest > bestValue:
            bestValue = previousBest                        

        i += 1

    return bestValue